pkg/service/providerindex/providerindex.go

```diff
@@ -34,28 +34,29 @@ type ProviderIndexService struct {
 	providerStore types.ProviderStore
 	findClient    ipnifind.Finder
 	publisher     publisher.Publisher
+	legacyClaims  LegacyClaimsFinder
 }
 
 var _ ProviderIndex = (*ProviderIndexService)(nil)
 
-// TBD access to legacy systems
-type LegacySystems interface{}
-
-func New(providerStore types.ProviderStore, findClient ipnifind.Finder, publisher publisher.Publisher, legacySystems LegacySystems) *ProviderIndexService {
+func New(providerStore types.ProviderStore, findClient ipnifind.Finder, publisher publisher.Publisher, legacyClaims LegacyClaimsFinder) *ProviderIndexService {
 	return &ProviderIndexService{
 		providerStore: providerStore,
 		findClient:    findClient,
 		publisher:     publisher,
+		legacyClaims:  legacyClaims,
 	}
 }
 
 // Find should do the following
 //  1. Read from the IPNI Storage cache to get a list of providers
-//     a. If there is no record in cache, query IPNI, filter out any non-content claims metadata, and store
+//     a. if there are no records in the cache, attempt to ready claims from legacy systems -- Dynamo tables & content
+//     claims storage, synthetically constructing provider results
+//     b. if no records are found in legacy systems, query IPNI, filter out any non-content claims metadata, and store
 //     the resulting records in the cache
-//     b. the are no records in the cache or IPNI, it can attempt to read from legacy systems -- Dynamo tables & content claims storage, synthetically constructing provider results
-//  2. With returned provider results, filter additionally for claim type. If space dids are set, calculate an encodedcontextid's by hashing space DID and Hash, and filter for a matching context id
-//     Future TODO: kick off a conversion task to update the recrds
+//  2. With returned provider results, filter additionally for claim type. If space dids are set, calculate an
+//     encodedcontextid's by hashing space DID and Hash, and filter for a matching context id
+//     Future TODO: kick off a conversion task to update the records
 func (pi *ProviderIndexService) Find(ctx context.Context, qk QueryKey) ([]model.ProviderResult, error) {
```

Comment:
that's interesting. The wrapper would try to find claims in the legacy storage and, if not found, call the current ProviderIndexService with the implementation it has today, is that correct? That wouldn't avoid having to synthetize `ProviderResult` records from the actual claims, as all the fetching and follow up queries are handled by `IndexingService`. Can you please elaborate on how it would improve testing?

An alternative would be for the legacy claims service to be a standalone service that `IndexingService` could ask for claims before going to `ProviderIndexService`. This has the drawback that results in legacy storage will not be cached, but I don't think this is so much of an issue beyond of the slight performance penalty it may have to query dynamodb instead of elasticcache. Another problem of this approach is that it will require duplicating the logic to fetch the indexes themselves.

I don't have a strong opinion, specially taking into account the fact that this code won't be there forever.

Reply:
> Can you please elaborate on how it would improve testing?

As far as I can tell, there's not actually a test yet that covers the change in this function. I just thought it might be easier to test by passing a mock that a wrapper could call and you could use to verify the fallback behaviour, than it might be to mock out all the dependencies of this `ProviderIndexService` and assert on the fact that it had been used.